### PR TITLE
[deckhouse-controller] forbide to change podSubnetNodeCIDRPrefix in running cluster

### PR DESCRIPTION
**Pull request by ViktorKram**
## Description
  Modified ClusterController validation webhook:
  1. Removed check Docker CRI compatibility with k8s 1.24 as docker is not supported at any point officially. 
  2. Removed nodegroups field from cluster configuration config.
  3. Added check podSubnetNodeCIDRPrefix, podSubnetCIDR, serviceSubnetCIDR cannot be changed in running cluster.

## Why do we need it, and what problem does it solve?
  Fix https://github.com/deckhouse/deckhouse/issues/5649. Deckhouse doesn't handle changes to `podSubnetNodeCIDRPrefix`, so we made it read-only field.

## What is the expected result?
  `podSubnetNodeCIDRPrefix`, `podSubnetCIDR`, `serviceSubnetCIDR` cannot be changed in a running cluster.

## Checklist
- [ ] The code is covered by unit tests.
- [ ] e2e tests passed.
- [x] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: deckhouse
type: fix
summary: Forbid changes to `podSubnetNodeCIDRPrefix` in a running cluster.
impact_level: low
```

### CODE REVIEW

File: modules/002-deckhouse/webhooks/validating/d8-cluster-configuration
Comment (name212) [inline modules/002-deckhouse/webhooks/validating/d8-cluster-configuration:1]: We can remove  endpoinds and nodegroup from config

**Review by name212 — CHANGES_REQUESTED** [review]

**Review by name212 — APPROVED** [review]

**Review by RomanenkoDenys — CHANGES_REQUESTED** [review]
some nits

**Review by RomanenkoDenys — APPROVED** [review]
LGTM

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @ViktorKram. I've started the workflow for label 'e2e/run/openstack'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @ViktorKram. I've started the workflow for label 'e2e/run/openstack'...

**Review by name212 — DISMISSED** [review]

**Comment by z9r5** [comment]
/e2e/abort openstack;Standard;containerd;1.25 6223859634 6223859634-1-con-1-25 /sys/deckhouse-oss/install:pr5792 debian@31.129.33.137

**Comment by z9r5** [comment]
@ViktorKram don't forget to delete failed clusters

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @z9r5. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @ViktorKram. I've started the workflow for label 'e2e/run/openstack'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @ViktorKram. I've started the workflow for label 'e2e/run/openstack'...

File: modules/002-deckhouse/webhooks/validating/d8-cluster-configuration [hunk 1]
```diff
@@ -72,45 +59,58 @@ EOF
     return 0
   fi
 
-  # defaultCRI cannot be changed if count of endpoints < 3
+  # podSubnetNodeCIDRPrefix cannot be changed in running cluster
```
Comment (nabokihms) [inline modules/002-deckhouse/webhooks/validating/d8-cluster-configuration:62]: You can reduce the boilerplate and make the hook more performant by storing the result on the `jq` in the intermediate objects:

```sh
oldConfig=$(context::jq -r '.review.request.oldObject.data."cluster-configuration.yaml"' | base64 -d)
newConfig=$(context::jq -r '.review.request.oldObject.data."cluster-configuration.yaml"' | base64 -d)
```

and then use them like:

```sh
oldPrefix=$(yq eval '.podSubnetNodeCIDRPrefix' - <<< "$oldConfig")
newPrefix=$(yq eval '.podSubnetNodeCIDRPrefix' - <<< "$newConfig")

```

File: modules/002-deckhouse/webhooks/validating/d8-cluster-configuration [hunk 2]
```diff
@@ -72,45 +59,58 @@ EOF
...
 EOF
       return 0
     fi
   fi
 
-  # check CRI compatibility with k8s 1.24, docker CRI was deprecated there
-  if context::jq -e -r '.review.request.operation == "UPDATE"' >/dev/null 2>&1; then
-    k8sVersion="$(context::jq -r '.review.request.object.data."cluster-configuration.yaml"' | base64 -d | yq eval '.kubernetesVersion // "1.21"' -)"
+  # podSubnetCIDR cannot be changed in running cluster
+    if context::jq -e -r '.review.request.operation == "UPDATE"' >/dev/null 2>&1; then
```
Comment (nabokihms) [inline modules/002-deckhouse/webhooks/validating/d8-cluster-configuration:76]: Let's reduce it to a single clause. The result should look like this:

```sh
if context::jq -e -r '.review.request.operation == "UPDATE"' >/dev/null 2>&1; then
  oldConfig=$(context::jq -r '.review.request.oldObject.data."cluster-configuration.yaml"' | base64 -d)
  newConfig=$(context::jq -r '.review.request.oldObject.data."cluster-configuration.yaml"' | base64 -d)

  oldPrefix=$(yq eval '.podSubnetNodeCIDRPrefix' - <<< "$oldConfig")
  newPrefix=$(yq eval '.podSubnetNodeCIDRPrefix' - <<< "$newConfig")
  if [ "${oldPrefix}" != "${newPrefix}" ]; then
    echo '{"allowed":false, "message":"it is forbidden to change podSubnetNodeCIDRPrefix in running cluster"}' > "$VALIDATING_RESPONSE_PATH"
    return 0
  fi

  oldPodCIDR=$(yq eval '.podSubnetCIDR' - <<< "$oldConfig")
  newPodCIDR=$(yq eval '.podSubnetCIDR' - <<< "$newConfig")
  if [ "${oldPodCIDR}" != "${newPodCIDR}" ]; then
    echo '{"allowed":false, "message":"it is forbidden to change podSubnetCIDR in running cluster"}' > "$VALIDATING_RESPONSE_PATH"
    return 0
  fi

  oldServiceCIDR=$(yq eval '.serviceSubnetCIDR' - <<< "$oldConfig")
  newServiceCIDR=$(yq eval '.serviceSubnetCIDR' - <<< "$newConfig")
  if [ "${oldServiceCIDR}" != "${newServiceCIDR}" ]; then
    echo '{"allowed":false, "message":"it is forbidden to change serviceSubnetCIDR in running cluster"}' > "$VALIDATING_RESPONSE_PATH"
    return 0
  fi
fi
```

File: modules/002-deckhouse/webhooks/validating/d8-cluster-configuration [hunk 3]
```diff
@@ -72,45 +59,58 @@ EOF
...
+      oldServiceSubnet="$(context::jq -r '.review.request.oldObject.data."cluster-configuration.yaml"' | base64 -d | yq eval '.serviceSubnetCIDR' -)"
+      newServiceSubnet="$(context::jq -r '.review.request.object.data."cluster-configuration.yaml"' | base64 -d | yq eval '.serviceSubnetCIDR' -)"
 
-        if [ "$criType" == "Docker" ]; then
-          cat <<EOF > "$VALIDATING_RESPONSE_PATH"
-{"allowed":false, "message":"Docker CRI is not supported by kubernetes 1.24. NodeGroup ${name} has incompatible CRI type"}
+      if [ "${oldServiceSubnet}" != "${newServiceSubnet}" ]; then
+        cat <<EOF > "$VALIDATING_RESPONSE_PATH"
+  {"allowed":false, "message":"it is forbidden to change serviceSubnetCIDR in running cluster"}
```
Comment (nabokihms) [inline modules/002-deckhouse/webhooks/validating/d8-cluster-configuration:95]:
```suggestion
  {"allowed":false, "message":"it is forbidden to change serviceSubnetCIDR in a running cluster"}
```

**Review by nabokihms — APPROVED** [review]

File: modules/002-deckhouse/webhooks/validating/d8-cluster-configuration [hunk 4]
```diff
@@ -66,56 +53,69 @@ EOF
...
 
-    if [ "$k8sVersion" == "1.24" ]; then
-      defaultCRI="$(context::jq -r '.review.request.object.data."cluster-configuration.yaml"' | base64 -d | yq eval '.defaultCRI // "Containerd"' -)"
+      # podSubnetNodeCIDRPrefix cannot be changed in a running cluster in a safe mode
+      oldPrefix=$(yq eval '.podSubnetNodeCIDRPrefix' - <<< "$oldConfig")
+      newPrefix=$(yq eval '.podSubnetNodeCIDRPrefix' - <<< "$newConfig")
+      if [ "${oldPrefix}" != "${newPrefix}" ]; then
+        echo '{"allowed":false, "message":"it is forbidden to change podSubnetNodeCIDRPrefix in a running cluster"}' > "$VALIDATING_RESPONSE_PATH"
+        return 0
+      fi
```
Comment (RomanenkoDenys) [inline modules/002-deckhouse/webhooks/validating/d8-cluster-configuration:88]: you eval oldPrefix and newPrefix 3 times. I think this is a copy/paste error

**Review by RomanenkoDenys — CHANGES_REQUESTED** [review]
some mistakes

File: modules/002-deckhouse/webhooks/validating/d8-cluster-configuration (hunk 4 above)
Comment (ViktorKram) [reply modules/002-deckhouse/webhooks/validating/d8-cluster-configuration]: Fixed, thanks.

**Pull request merged**